### PR TITLE
Wrappers

### PR DESCRIPTION
**Pull request by Zeus3101**
Solves #125 . Let me know if this structure is fine. I've based it off TF Agents.

### CODE REVIEW

**Comment by Zeus3101** [comment]
Will depend on how VecEnvs are implemented too. I want to add some functions like get_state, set_state etc too. tf agents has defined specs for every different kind of space which is too complex to implement at this stage. Are there any other specific features you'd want? @Sharad24 @ajaysub110 @Het-Shah

File: genrl/environments/gym_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    :type env: string
+    :type n_envs: None, int
+    """
+    def __init__(self, env, n_envs=None):
+        if n_envs is None:
+            self._vec = False
+            self._env = gym.make(env)
+        else:
+            self._vec = True
+            self._env = venv(env, n_envs)
```
Comment (Sharad24) [inline genrl/environments/gym_wrapper.py:25]: parallel arg is missing here

Comment (Zeus3101) [reply genrl/environments/gym_wrapper.py:25]: Noted

**Comment by Zeus3101** [comment]
Here's a rough list of things I'll implement. I'll change the structure a little bit later on.

- [ ] `BaseWrapper` class for all envs
- [ ] `GymWrapper` class for all gym envs
       - Enable Trainer to use the same syntax with or without VecEnvs
       - Methods to get/set current state, get current timestep, done, info
- [ ] `AtariPreprocessing` class to help preprocess input from Atari envs
       - Grayscale
       - Frameskip
       - Max Pooling over last two observations
       - Resizing to (84, 84)
- [ ] `FrameStack` class to stack last 4 observations
- [ ] `FireReset` class for those Atari envs where you need to perform some fixed action before it starts training
- [ ] `NoopReset` class for those Atari envs that always reset to same initial state
- [ ] `Atari` class to stack all relevant wrappers into one single wrapper we'll use in the Trainer class

**Comment by Zeus3101** [comment]
@Sharad24 check out the last commit and tell me if the changes are fine. It would be better if you did it in your PR, I just put it up here as an example.

File: genrl/environments/vectorised_envs.py [hunk 2]
```diff
@@ -46,6 +46,8 @@ class VecEnv(ABC):
     def __init__(self, env, n_envs=2):
         self.envs = create_envs(env, n_envs)
         self._n_envs = len(self.envs)
+        self.observation_space = self.env.observation_space
+        self.action_space = self.env.action_space
```
Comment (Het-Shah) [inline genrl/environments/vectorised_envs.py:50]: why do we need this? I mean I already have it in worker method in get_spaces.

Comment (Zeus3101) [reply genrl/environments/vectorised_envs.py:50]: This is supposed to work for both regular envs and vectorised envs. So if we use this as a wrapper, we need to be directly be able to call `env.observation_space` like you would call normally. My aim was not to affect normal usage.

**Comment by Zeus3101** [comment]
Update: (not yet ready to merge but do review)

- `GymEnv` is a general wrapper function to use for any agent. No speedup in performance
- `AtariEnv` is a wrapper function for Atari envs which can only be used for DQN atm. Will try to expand to other agents too. This increases performance speed a little bit. 
- Still need to test it for all algorithms and add a few additional functions.

It'll be ready to merge by Monday.

File: tests/test_environments/test_atari_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def test_atari_env(self):
+        """
+        Tests working of Atari Wrappers and the AtariEnv function
+        """
+        env = AtariEnv("Pong-v0")
+        algo = DQN("cnn", env, noisy_dqn=True, prioritized_replay=True)
+
+        trainer = OffPolicyTrainer(
+            algo, env, log_mode=["csv"], logdir="./logs", epochs=1, steps_per_epoch=200
```
Comment (Sharad24) [inline tests/test_environments/test_atari_wrappers.py:52]: Does this work?

**Comment by Sharad24** [comment]
Looks good to me till now

File: tests/test_environments/test_atari_wrappers.py (hunk 3 above)
Comment (Zeus3101) [reply tests/test_environments/test_atari_wrappers.py:52]: Yes

**Comment by Sharad24** [comment]
Three more left, is it?

**Comment by Zeus3101** [comment]
Actually, I'm done locally with the `FireReset` class. There's only support for VecEnvs left

**Comment by Sharad24** [comment]
@Zeus3101 What about input Normalization based wrappers? They are pretty important for envs like MountainCar, Pybullet/Mujoco + some atari envs which dont train without them at all

**Comment by Zeus3101** [comment]
Ok I'll work on those too, any others you can think of that we need?

**Comment by Sharad24** [comment]
Code coverage should ideally increase a lot here. Because some tests were commented as part of #111

**Comment by Zeus3101** [comment]
> Code coverage should ideally increase a lot here. Because some tests were commented as part of #111

I'll sort these out soon.

**Comment by Zeus3101** [comment]
A few changes I'm making to the VecEnvs part:
- Moved the file to the environments folder
- Renamed to `vec_envs.py` because it stood out too much with the capitalisation (no other file is capitalised)
- Removed `create_envs` and `venv` from the file and added the same functionality to `suite.py` which will contain all the wrappers. Now, the user can use:
`env = VectorEnv("Pong-v0", "atari", n_envs=3, parallel=False)`
- The `VecEnv` base class now takes in argument `envs` which is passed as a list of envs by the function `VectorEnv`

**Comment by Sharad24** [comment]
Ok, will review it in a few hours.

**Comment by Zeus3101** [comment]
The current build's failing because of #158

**Comment by Sharad24** [comment]
Will resolve that

**Comment by Sharad24** [comment]
Tests pass, you need to reformat your files now I guess. FYI, the pre-commit is exactly for this purpose. It automatically runs black when you run `git commit ...`

`pip install pre-commit
`pre-commit run --all-files`

**Comment by Sharad24** [comment]
How much left to do here?

**Comment by Sharad24** [comment]
Link check fails

**Comment by Sharad24** [comment]
Do you want me to fix the pre-commit for you? It will only take 30 seconds for me

**Comment by Zeus3101** [comment]
For now, I think it would be best to merge this after I get tests working and everything functional. I only have to implement Atari support for Actor Critic algos and that's taking some time as there's a lot of small changes that need to be done. I'll handle them in a separate PR after this gets merged.

**Comment by Zeus3101** [comment]
> Do you want me to fix the pre-commit for you? It will only take like 30 seconds

Sure!

**Comment by Sharad24** [comment]
What all is left? 
Can you try to get rid of the code climates issues?

Just as a reminder, the longer we dont merge, the more work for you in resolving conflicts, adapting to new code, etc

**Comment by Zeus3101** [comment]
Right now, I need to write some more tests. A major problem is I cant get parallel VecEnvs working properly.

As for the code climate issues, by using wrappers, some of them can't be avoided (multiple wrappers on the same env is causing core methods like `step()`. I'll try to resolve them where possible.

**Comment by Zeus3101** [comment]
I'll try to finish it all by tomorrow. This is a lot of new code for GenRL so I just want to make sure everything works properly before we merge.

**Comment by Sharad24** [comment]
Sure, sounds great!

**Comment by Zeus3101** [comment]
Ready to merge. Will track other issues separately. Namely #166 and #172

**Comment by Sharad24** [comment]
Code climate 8 issues? We need to fix these.

**Comment by Zeus3101** [comment]
For the code climate issues, one way to solve some of them would be to decrease the customizability for the user by having lesser variables for them to edit. 

Pro: The code would be far neater then.
Con: They would have to define their own function to apply wrappers if they want customizability. Maybe we could attach a `tutorial.md` to help write customized wrappers?

**Comment by Zeus3101** [comment]
This is ready to merge. 1 codeclimate issue will be resolved in #170 . The others seem unavoidable. Take a look and let me know if any other changes are required. @Sharad24 @Het-Shah @ajaysub110

**Comment by Zeus3101** [comment]
Without the `vec_` it would be quite confusing. Since we're using relative imports, some files would end up having `from .wrappers import` and that doesn't make sense. Either that or we can do away with the `vec_env` folder entirely.

**Comment by Sharad24** [comment]
Okay cool.

**Comment by Sharad24** [comment]
Creating an issue #174
